modules/standard/CTypes.chpl

```diff
@@ -0,0 +1,726 @@
+    :returns: a c_ptr(eltType) to allocated memory
+    */
+  inline proc c_calloc(type eltType, size: integral) : c_ptr(eltType) {
+    const alloc_size = size.safeCast(size_t) * c_sizeof(eltType);
+    return chpl_here_calloc(alloc_size, 1, offset_ARRAY_ELEMENTS):c_ptr(eltType);
+  }
+
+  /*
+    Allocate memory that is not initialized. This memory should eventually be
+    freed with :proc:`c_free`.
+
+    :arg eltType: the type of the elements to allocate
+    :arg size: the number of elements to allocate space for
+    :returns: a c_ptr(eltType) to allocated memory
+    */
+  inline proc c_malloc(type eltType, size: integral) : c_ptr(eltType) {
+    const alloc_size = size.safeCast(size_t) * c_sizeof(eltType);
+    return chpl_here_alloc(alloc_size, offset_ARRAY_ELEMENTS):c_ptr(eltType);
+  }
+
+  /*
+    Allocate aligned memory that is not initialized. This memory
+    should be eventually freed with :proc:`c_free`.
+
+    This function is intended to behave similarly to the C17
+    function aligned_alloc.
+
+    :arg eltType: the type of the elements to allocate
+    :arg alignment: the memory alignment of the allocation
+                    which must be a power of two and a multiple
+                    of ``c_sizeof(c_void_ptr)``.
+    :arg size: the number of elements to allocate space for
+    :returns: a ``c_ptr(eltType)`` to allocated memory
+    */
+  inline proc c_aligned_alloc(type eltType,
+                              alignment : integral,
+                              size: integral) : c_ptr(eltType) {
+    // check alignment, size restriction
+    if boundsChecking {
+      var one:size_t = 1;
+      // Round the alignment up to the nearest power of 2
+      var aln = alignment.safeCast(size_t);
+      if aln == 0 then
+        halt("c_aligned_alloc called with alignment of 0");
+      var p = log2(aln); // power of 2 rounded down
+      // compute alignment rounded up
+      if (one << p) < aln then
+        p += 1;
+      assert(aln <= (one << p));
+      if aln != (one << p) then
+        halt("c_aligned_alloc called with non-power-of-2 alignment ", aln);
+      if alignment < c_sizeof(c_void_ptr) then
+        halt("c_aligned_alloc called with alignment smaller than pointer size");
+    }
+
+    const alloc_size = size.safeCast(size_t) * c_sizeof(eltType);
+    return chpl_here_aligned_alloc(alignment.safeCast(size_t),
+                                   alloc_size,
+                                   offset_ARRAY_ELEMENTS):c_ptr(eltType);
+  }
+
+  /* Free memory that was allocated with :proc:`c_calloc` or :proc:`c_malloc`.
+
+    :arg data: the c_ptr to memory that was allocated. Note that both
+               `c_ptr(t)` and `c_void_ptr` can be passed to this argument.
+    */
+  inline proc c_free(data: c_void_ptr) {
+    chpl_here_free(data);
+  }
+
+  /* Returns true if t is a c_ptr type or c_void_ptr.
+   */
+  proc isAnyCPtr(type t:c_ptr) param return true;
```

Comment:
Could you add a Future Work to the PR message about checking if we are happy with the name of this function? (or if it is in some issue elsewhere a link to that?)

Reply:
Done.